Make setup.py pip compatible

bergmannf opened the pull request:
Fix the documentation generation code to allow installation using `pip`.

`pip install git+https://github.com/openSUSE/osc2.git` does not use the module's directory while running, so the `pydoc` command can not find the modules.

marcus-h commented:
First of all, thanks for the PR!

On 2017-12-14 12:05:06 +0000, Florian Bergmann wrote:
> Fix the documentation generation code to allow installation using `pip`.
> 
> `pip install git+https://github.com/openSUSE/osc2.git` does not use the module's directory while running, so the `pydoc` command can not find the modules.

Do you know how pip calls the setup.py script?

I just had a quick glance at your patches:

> diff --git a/setup.py b/setup.py
> index 550f276..2877e8c 100755
> --- a/setup.py
> +++ b/setup.py
> @@ -1,15 +1,18 @@
>  #!/usr/bin/python
>  
> -import os
> -from setuptools import setup, find_packages
>  from distutils.command.build import build
>  from pydoc import writedoc
>  
> +import os
> +from setuptools import setup
> +
>  
Hmm unnecessary "patch noise": removing the find_packages import is ok,
but moving the imports around is not needed.

> @@ -49,4 +52,11 @@ def run(self, *args, **kwargs):
>        test_suite='test.suite',
>        entry_points={'console_scripts': ['osc2 = osc2.cli.cli:main']},
>        license='GPL',
> -      cmdclass={'build': BuildDocumentation})
> +      cmdclass={'build': BuildDocumentation},
> +      install_requires=[
> +          "lxml",
> +          "pycurl",

Hmm osc2 does not use pycurl.

> +          "urlgrabber",
> +          "jinja2"
> +      ],
> +      classifiers=["Programming Language :: Python :: 2"])
> 
Hmm I'm not sure if I got this classifiers thing right, but isn't this
classifier too "general"? Maybe rather something like
["Programming Language :: Python :: 2.7"]
?

> diff --git a/setup.py b/setup.py
> index 2877e8c..78d9184 100755
> --- a/setup.py
> +++ b/setup.py

> +import sys
>  from distutils.command.build import build
>  from pydoc import writedoc
>  
> @@ -12,7 +13,10 @@ def _generate_docs(self):
>          # FIXME: this is really hacky
>          # Need to work in the modules directory.
>          # Otherwise install with e.g. pip will not work.
> -        olddir = __file__
> +        olddir = os.getcwd()
> +        module_dir = os.path.dirname(os.path.abspath(__file__))
> +        sys.path.append(module_dir)

Is this really needed?

> +        os.chdir(module_dir)
>          html_dir = os.path.join('docs', 'html')
>          if not os.path.exists(html_dir):
>              os.makedirs(html_dir)
> 

Can you squash both commits into one?

bergmannf commented:
Thanks for the review! I went over your recommendations:

> Do you know how pip calls the setup.py script?

Unfortunately I don't know how. I just tested it by installing via `pip` and seeing if it works for now.

> Hmm unnecessary "patch noise" [...]

Fixed the import reordering - something my editor does automatically, sorry about that.

> Hmm osc2 does not use pycurl.

True that should be resolved transitively from `urlgrabber`, so I removed it.

> [...] Maybe rather something like ["Programming Language :: Python :: 2.7"]

The classifiers are used for searching only, but I fixed it to `Python :: 2.7` ([more about classifiers](https://packaging.python.org/tutorials/distributing-packages/#classifiers))

I tested it in a virtual environment and without adding `sys.path.append(..)` and it would fail with:

```
    no Python documentation found for 'osc2'
    error: [Errno 2] No such file or directory
```

So unfortunately it seems needed.

Also squashed everything down into one commit.

marcus-h commented:
On 2017-12-15 09:33:37 +0000, Florian Bergmann wrote:
> > [...] Maybe rather something like ["Programming Language :: Python :: 2.7"]
> 
> The classifiers are used for searching only, but I fixed it to `Python :: 2.7` ([more about classifiers](https://packaging.python.org/tutorials/distributing-packages/#classifiers))
> 
Thanks.

> I tested it in a virtual environment and without adding `sys.path.append(..)` and it would fail with:
> 
> ```
>     no Python documentation found for 'osc2'
>     error: [Errno 2] No such file or directory
> ```
> 
> So unfortunately it seems needed.
> 
> Also squashed everything down into one commit.
> 
Merged. Thanks a lot!
